README.md
# DD-Guard (Diabetes Data Guard)

## What it is

DD-Guard provides real time blood glucose and pump data for the "*[Medtronic Minimed 670G](https://www.medtronicdiabetes.com/products/minimed-670g-insulin-pump-system)*"  insulin pump system directly to your smart phone. It intends to be simple to use and easy to setup.

![ddguard-app-screen](img/ddguard-app-screen.png)

The 670G pump features continuous blood glucose measurements (CGM) via the "*Guardian Link 3*" sensor and stores the data on the device. The data can be viewed on the local display. However to date Medtronic provides no means of displaying the 670Gs real time data remotely on a mobile device. And that's exactly the functionality that DD-Guard adds to the system, so care givers can watch immediately the most important data from the sensor and the pump conveniently on their smart phones, wherever they are.

DD-Guard was inspired by the [NightScout](http://www.nightscout.info) project and the #WeAreNotWaiting community which promotes DIY efforts to take advantage of the latest technology to make life easier for people with Type-1 Diabetes and their care givers.



## Project Status

My wife and I have been continuously using DD-Guard for a few months in the real world to monitor my daughters blood glucose level and pump status. On the app screen of the mobile phone we get updated data for blood glucose level (including history graph) and trend, active insulin, remaining insulin units in the pumps tank and both pump and sensor battery status (displayed alternatively). Bolus events are displayed on the graph. The displayed data is color coded according to the actual conditions so it is immediately clear if there is anything critical which needs to be acted upon. The system is robust and running for weeks without any issues. 

It is possible for my daughter to take the small gateway device with her when she is going to spend the night at a friends house, so we can still monitor her data. The gateway works as long as it has a power supply and a Wifi network connection.

The gateway software now runs reliably also on the tiny [Raspberry Pi Zero](https://www.raspberrypi.org/products/raspberry-pi-zero-w/) device which can be powered for may hours from a battery pack and therefore carried in a small backpack, making the gateway truly mobile.

A Nightscout uploader option is available for the DD-Guard gateway, so it can upload the live sensor and pump data also to a Nightscout server and replace the commonly used Android phone uploader for this system (unless you need all the bells and whistles from the latter).



## Screenshots

These are some typical screenshots from the smartphone app.

![Screenshots](img/ddguard-screenshots.png)



## How it works

The basic idea is to receive the real time data which was collected by the 670G with the DD-Guard gateway via the "*Contour Next Link* 2.4" glucose meter which operates as the radio bridge and then uploads the data to the cloud and a mobile device where it is eventually displayed with the DD-Guard app.

The DD-Guard gateway is a small single board computer, like the Raspberry Pi where the radio bridge is plugged into one of its USB ports. Cloud connection is established via the gateways Wifi.

![ddguard-overview](img/ddguard-overview.png)



## What hardware do you need

In order to use DD-Guard you need the following items:

- Medtronic Minimed 670G insulin pump (also works with 640G)
- Guardian Link blood glucose sensor and radio transmitter 
- Contour Next Link 2.4 blood glucose meter and radio bridge
- A single board computer with USB and Wifi like RaspberryPi 3 or similar as DD-Guard gateway
- A smartphone

If the person you build this system for is a T1D patient on insulin pump therapy you probably already have the first 3 items if you chose the Medtronic device. And chances are good you already have a smartphone.

So all you need to do is build your own gateway. It needs a USB port to connect to the radio bridge and Wifi to connect to the cloud.



![rpi-zero-cnl24](img/rpi-zero-cnl24.jpg)

## What software do you need

These are the logical software components which are needed to make it all work together:

- The **DD-Guard smartphone app** which receives the data from the cloud and displays it.
- The **Cloud service** which receives the data from the gateway and forwards it to the smartphone app
- The **DD-Guard gateway software** which periodically receives the data from the pump and uploads it to the cloud service



## Disclaimer

This project is not associated to or endorsed by [Medtronic](https://www.medtronicdiabetes.com). If you decide to use DD-Guard then you do this entirely at your own risk. I am not reliable for any damage it might cause. 


HISTORY.md
# VERSION HISTORY

### 0.1 [26/09/2019]

First public release

### 0.2 [24/10/2019]

- Add handling of parameters from configuration file
- Add handling of BGL status codes
- Add handling of display colors according to limits

### 0.3 [03/11/2019]

- Run timer function as asynchronous thread to avoid cloud disconnection
- Add install script

### 0.4 [13/11/2019]

* Add missing sensor exception codes
* Create cnl24driverlib
* Integrate Nightscout uploader

###  0.5 [03/03/2020]

* Improved robustness of CNL2.4 driver

### 0.6 [16/04/2020]

* Parse all data in PumpStatusResponseMessage of CNL2.4 driver
* Add more status data to Blynk uploader

### 0.7 [10/08/2020]
* Add pump status handling to Nightscout lib
* Python3 support

### 0.8 [23/11/2020]

* Account for pump time drift

* Sync pump status and sensor data download with BGL timestamp for faster update

  

sensor_codes.py

class SENSOR_EXCEPTIONS:
   SENSOR_OK                   = 0x0300
   SENSOR_OK_STR               = "Sensor OK"
   SENSOR_INIT                 = 0x0301
   SENSOR_INIT_STR             = "Sensor warming up"
   SENSOR_CAL_NEEDED           = 0x0302
   SENSOR_CAL_NEEDED_STR       = "Calibrate sensor now "
   SENSOR_ERROR                = 0x0303
   SENSOR_ERROR_STR            = "Updating sensor"
   SENSOR_CAL_ERROR            = 0x0304
   SENSOR_CAL_ERROR_STR        = "Calibration error"
   SENSOR_CHANGE_SENSOR        = 0x0305
   SENSOR_CHANGE_SENSOR_STR    = "Change sensor"
   SENSOR_END_OF_LIFE          = 0x0306
   SENSOR_END_OF_LIFE_STR      = "Sensor expired"
   SENSOR_NOT_READY            = 0x0307
   SENSOR_NOT_READY_STR        = "Sensor not ready"
   SENSOR_READING_HIGH         = 0x0308
   SENSOR_READING_HIGH_STR     = "Sensor reading too high"
   SENSOR_READING_LOW          = 0x0309
   SENSOR_READING_LOW_STR      = "Sensor reading too low"
   SENSOR_CAL_PENDING          = 0x030A
   SENSOR_CAL_PENDING_STR      = "Calibrating sensor"
   SENSOR_CHANGE_CAL_ERROR     = 0x030B
   SENSOR_CHANGE_CAL_ERROR_STR = "Cal error - Change sensor"
   SENSOR_TIME_UNKNOWN         = 0x030C
   SENSOR_TIME_UNKNOWN_STR     = "Time unknown"
   SENSOR_LOST                 = 0
   SENSOR_LOST_STR             = "Lost connection to sensor"


nightscoutlib.py
###############################################################################
#  
#  Diabetes Data Guard (DD-Guard): Nightscout uploader library
#  
#  Description:
#
#    This library implements the uploader of the live sensor and pump  
#    data to the Nightscout REST API
#  
#  Author:
#
#    Ondrej Wisniewski (ondrej.wisniewski *at* gmail.com)
#  
#  Changelog:
#
#    19/11/2019 - Initial public release
#    25/11/2019 - Add exception code handling
#    29/11/2019 - Upload of real device serial and pump date
#    13/12/2019 - Check for "lost sensor" condition
#    14/04/2020 - Adapt to new data format from CNL driver
#    27/04/2020 - Add pump status handling
#    28/06/2020 - Syntax updates for Python3
#
#  Copyright 2019-2020, Ondrej Wisniewski 
#  
#  This file is part of the DD-Guard project.
#  
#  DD-Guard is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
# 
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
# 
#  You should have received a copy of the GNU General Public License
#  along with crelay.  If not, see <http://www.gnu.org/licenses/>.
#  
###############################################################################
import json
import syslog
import hashlib
import requests
from sensor_codes import SENSOR_EXCEPTIONS


# Nightscout error codes
class NS_ERROR:
   SENSOR_NOT_ACTIVE     = 1  # "?SN"
   MINIMAL_DEVIATION     = 2  # "?MD"
   NO_ANTENNA            = 3  # "?NA"
   SENSOR_NOT_CALIBRATED = 5  # "?NC"
   COUNTS_DEVIATION      = 6  # "?CD"
   ABSOLUTE_DEVIATION    = 9  # "?AD"
   POWER_DEVIATION       = 10 # "???"
   BAD_RF                = 12 # "?RF"

# Nightscout trend codes
class NS_TREND:
   NONE                  = "NONE"
   TRIPLE_UP             = "TripleUp"
   DOUBLE_UP             = "DoubleUp"
   SINGLE_UP             = "SingleUp"
   FOURTY_FIVE_UP        = "FortyFiveUp"
   FLAT                  = "Flat"
   FOURTY_FIVE_DOWN      = "FortyFiveDown"
   SINGLE_DOWN           = "SingleDown"
   DOUBLE_DOWN           = "DoubleDown"
   TRIPLE_DOWN           = "TripleDown"
   NOT_COMPUTABLE        = "NOT COMPUTABLE"
   RATE_OUT_OF_RANGE     = "RATE OUT OF RANGE"
   NOT_SET               = "NONE"

# Nightscout uploader class
class nightscout_uploader(object):
   
   def __init__(self, server, secret):
      if "http" in server.strip():
         self.ns_url  = server.strip()
      else:
         self.ns_url     = "http://"+server.strip()
      self.api_secret = hashlib.sha1(str(secret.strip()).encode('utf-8')).hexdigest()
      self.api_base   = "/api/v1/"
      self.device     = "medtronic-600://"
      self.headers    = {
                           "user-agent":"dd-guard",
                           "Content-Type":"application/json",
                           "api-secret":self.api_secret
                        }
      
   # Trend mapping
   def direction_str(self, trend):
      if trend   == -3:
         return NS_TREND.TRIPLE_DOWN
      elif trend == -2:
         return NS_TREND.DOUBLE_DOWN
      elif trend == -1:
         return NS_TREND.SINGLE_DOWN
      elif trend == 0:
         return NS_TREND.FLAT
      elif trend == 1:
         return NS_TREND.SINGLE_UP
      elif trend == 2:
         return NS_TREND.DOUBLE_UP
      elif trend == 3:
         return NS_TREND.TRIPLE_UP
      else:
         return NS_TREND.RATE_OUT_OF_RANGE
      
   # Exception code mapping
   def exception_code(self, sgv):
      if sgv in [SENSOR_EXCEPTIONS.SENSOR_CAL_NEEDED]:
         return NS_ERROR.SENSOR_NOT_CALIBRATED,NS_TREND.NOT_COMPUTABLE 
      elif sgv in [SENSOR_EXCEPTIONS.SENSOR_CHANGE_CAL_ERROR, 
                   SENSOR_EXCEPTIONS.SENSOR_CHANGE_SENSOR,
                   SENSOR_EXCEPTIONS.SENSOR_END_OF_LIFE]:
         return NS_ERROR.SENSOR_NOT_ACTIVE,NS_TREND.NOT_COMPUTABLE
      elif sgv in [SENSOR_EXCEPTIONS.SENSOR_READING_LOW]:
         return 40,NS_TREND.RATE_OUT_OF_RANGE
      elif sgv in [SENSOR_EXCEPTIONS.SENSOR_READING_HIGH]:
         return 400,NS_TREND.RATE_OUT_OF_RANGE
      elif sgv in [SENSOR_EXCEPTIONS.SENSOR_CAL_PENDING, 
                   SENSOR_EXCEPTIONS.SENSOR_INIT, 
                   SENSOR_EXCEPTIONS.SENSOR_TIME_UNKNOWN, 
                   SENSOR_EXCEPTIONS.SENSOR_NOT_READY, 
                   SENSOR_EXCEPTIONS.SENSOR_ERROR]:
         return NS_ERROR.NO_ANTENNA,NS_TREND.NOT_COMPUTABLE
      
      
   #########################################################
   #
   # Function:    upload_entries()
   # Description: Upload sensor data via the entries/ 
   #              API endpoint
   # 
   #########################################################
   def upload_entries(self, data):

      rc = True
      url = self.ns_url + self.api_base + "entries.json"
      sgv = data["sensorBGL"]
      trend = data["trendArrow"]
      date = data["sensorBGLTimestamp"]
      
      # Check for "lost sensor" condition
      # We don't upload any sensor data in this case
      if (sgv == 0) and (trend == -3): # and (date.strftime("%c").find("01:00:00 1970") != -1):
         print("Sensor lost, not uploading SGV data")
         return False
      
      # Check for exception codes
      if sgv >= 0x0300:
         sgv,trend_str = self.exception_code(sgv)
      else:
         trend_str = self.direction_str(trend)
      
      payload = {
            "device":self.device+data["serial"],
            "type":"sgv",
            "dateString":date.isoformat(),
            "date":int(date.strftime("%s"))*1000,
            "sgv":sgv,
            "direction":trend_str
         }
      #print("url: " + url)
      #print("headers: "+json.dumps(self.headers))
      #print("payload: "+json.dumps(payload))
      
      try:
         #print "Send API request"
         r = requests.post(url, headers = self.headers, data = json.dumps(payload))
         #print "API response: "+r.text
         if r.status_code != requests.codes.ok:
            syslog.syslog(syslog.LOG_ERR, "Uploading entries record returned error "+str(r.status_code))
            rc = False
      except:
         #print "Uploading entries record failed with exception"
         syslog.syslog(syslog.LOG_ERR, "Uploading entries record failed with exception")
         rc = False
   
      return rc


   #########################################################
   #
   # Function:    upload_devicestatus()
   # Description: Upload pump data via the devicestatus/ 
   #              API endpoint
   # 
   #########################################################
   def upload_devicestatus(self, data):
   
      rc = True
      url = self.ns_url + self.api_base + "devicestatus.json"
      date = data["pumpTime"]
      
      # Build pump status
      if data["pumpStatus"]["cgmActive"]:
         status = " | "
         if data["sensorStatus"]["exception"]==0:
            status = status + str(data["sensorBatteryLevelPercentage"]) + "% "
         if data["sensorCalMinutesRemaining"]>0:
            status = status + "{0}:{1:02d}h".format(int(data["sensorCalMinutesRemaining"]/60),data["sensorCalMinutesRemaining"]%60)
      else:
         status = ""
      
      payload = {
            "device":self.device+data["serial"],
            "created_at": int(date.strftime("%s"))*1000,
            "uploaderBattery":100, # FIXME
            "pump": {
               "clock":int(date.strftime("%s"))*1000,
               "reservoir":data["insulinUnitsRemaining"],
               "battery": {
                  "percent":data["batteryLevelPercentage"]
               },
               "iob": {
                  "timestamp":int(date.strftime("%s"))*1000,
                  "bolusiob":data["activeInsulin"]
               },
               "status": {
                  "bolusing":True if (data["pumpStatus"]["bolusingNormal"]|data["pumpStatus"]["bolusingSquare"]|data["pumpStatus"]["bolusingDual"]) else False,
                  "suspended":True if data["pumpStatus"]["suspended"] else False,
                  "status":status
               }
            }
         }
      #print "url: " + url
      #print "payload: "+json.dumps(payload)
  
      try:
         #print "Send API request"
         r = requests.post(url, headers = self.headers, data = json.dumps(payload))
         #print "API response: "+r.text
         if r.status_code != requests.codes.ok:
            syslog.syslog(syslog.LOG_ERR, "Uploading entries record returned error "+str(r.status_code))
            rc = False
      except:
         #print "Uploading entries record failed with exception"
         syslog.syslog(syslog.LOG_ERR, "Uploading entries record failed with exception")
         rc = False
   
      # TODO: delete old entries
   
      return rc
   

   #########################################################
   #
   # Function:    upload()
   # Description: Upload sensor and pump data to the 
   #              Nightscout REST API
   # 
   #########################################################
   def upload(self, data):
   
      rc = True
      if data != None:
         print("Uploading data to Nightscout")
         
         # Upload sensor data
         rc = self.upload_entries(data)
   
         # Upload pump data
         rc &= self.upload_devicestatus(data)

      return rc   


ddguard.py
#!/usr/bin/env python3
###############################################################################
#  
#  Diabetes Data Guard (DD-Guard): Gateway module
#  
#  Description:
#
#    The DD-Guard gateway module periodically receives real time data from the 
#    Medtronic Minimed 670G insulin pump and uploads it to the cloud service
#  
#  Dependencies:
#
#    DD-Guard uses the Python driver by paazan for the "Contour Next Link 2.4" 
#    radio bridge to the Minimed 670G to download real time data from the pump.
#    https://github.com/pazaan/decoding-contour-next-link
#    
#    For the cloud connection and app communication the official Blynk Python
#    library is used.
#    https://github.com/blynkkk/lib-python
#
#  Author:
#
#    Ondrej Wisniewski (ondrej.wisniewski *at* gmail.com)
#  
#  Changelog:
#
#    23/09/2019 - Initial public release
#    13/10/2019 - Add handling of parameters from configuration file
#    24/10/2019 - Add handling of BGL status codes
#    24/10/2019 - Add handling of display colors according to limits
#    02/11/2019 - Run timer function as asynchronous thread
#    07/11/2019 - Add missing sensor exception codes
#    24/11/2019 - Integrate Nightscout uploader
#    03/12/2019 - Adapt to modified library names
#    03/12/2019 - Make Blynk uploader optional
#    11/02/2020 - Add Blynk virtual pin for active insulin
#    03/03/2020 - Improved robustness of CNL2.4 driver
#    14/04/2020 - Adapt to new data format from CNL driver
#    15/04/2020 - Add more status data to blynk uploader
#    28/06/2020 - Syntax updates for Python3
#    09/11/2020 - Replace Blynk timer with Python timer,
#                 Account for Pump Time drift
#
#  TODO:
#    - Upload missed data when the pump returns into range
#    - Add some notification mechanism for alarms e.g. Telegram or Pushover message
#    - Upload data to Tidepool
#
#  Copyright 2019-2020, Ondrej Wisniewski 
#  
#  This file is part of the DD-Guard project.
#  
#  DD-Guard is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
# 
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
# 
#  You should have received a copy of the GNU General Public License
#  along with crelay.  If not, see <http://www.gnu.org/licenses/>.
#  
###############################################################################

import blynklib
import signal
import syslog
import sys
import time
import threading 
if sys.version_info[0] < 3:
    from ConfigParser import ConfigParser
else:
    from configparser import ConfigParser
import datetime
import cnl24driverlib
import nightscoutlib
from sensor_codes import SENSOR_EXCEPTIONS

VERSION = "0.8"

UPDATE_INTERVAL = 300
RETRY_INTERVAL  = 180
RETRY_DELAY     = 5
MAX_RETRIES_AT_FAILURE = 3

# virtual pin definitions
VPIN_SENSOR  = 1
VPIN_BATTERY = 2
VPIN_UNITS   = 3
VPIN_ARROWS  = 4
VPIN_STATUS  = 5
VPIN_ACTINS  = 6
VPIN_LASTBOLUS = 7

# color definitions
BLYNK_WHITE  = "#F0F0F0"
BLYNK_GREEN  = "#23C48E"
BLYNK_BLUE   = "#04C0F8"
BLYNK_YELLOW = "#ED9D00"
BLYNK_RED    = "#D3435C"
BLYNK_DARK_BLUE = "#5F7CD8"

sensor_exception_codes = {
    SENSOR_EXCEPTIONS.SENSOR_OK:               SENSOR_EXCEPTIONS.SENSOR_OK_STR,
    SENSOR_EXCEPTIONS.SENSOR_INIT:             SENSOR_EXCEPTIONS.SENSOR_INIT_STR,
    SENSOR_EXCEPTIONS.SENSOR_CAL_NEEDED:       SENSOR_EXCEPTIONS.SENSOR_CAL_NEEDED_STR,
    SENSOR_EXCEPTIONS.SENSOR_ERROR:            SENSOR_EXCEPTIONS.SENSOR_ERROR_STR,
    SENSOR_EXCEPTIONS.SENSOR_CAL_ERROR:        SENSOR_EXCEPTIONS.SENSOR_CAL_ERROR_STR,
    SENSOR_EXCEPTIONS.SENSOR_CHANGE_SENSOR:    SENSOR_EXCEPTIONS.SENSOR_CHANGE_SENSOR_STR,
    SENSOR_EXCEPTIONS.SENSOR_END_OF_LIFE:      SENSOR_EXCEPTIONS.SENSOR_END_OF_LIFE_STR,
    SENSOR_EXCEPTIONS.SENSOR_NOT_READY:        SENSOR_EXCEPTIONS.SENSOR_NOT_READY_STR,
    SENSOR_EXCEPTIONS.SENSOR_READING_HIGH:     SENSOR_EXCEPTIONS.SENSOR_READING_HIGH_STR,
    SENSOR_EXCEPTIONS.SENSOR_READING_LOW:      SENSOR_EXCEPTIONS.SENSOR_READING_LOW_STR,
    SENSOR_EXCEPTIONS.SENSOR_CAL_PENDING:      SENSOR_EXCEPTIONS.SENSOR_CAL_PENDING_STR,
    SENSOR_EXCEPTIONS.SENSOR_CHANGE_CAL_ERROR: SENSOR_EXCEPTIONS.SENSOR_CHANGE_CAL_ERROR_STR,
    SENSOR_EXCEPTIONS.SENSOR_TIME_UNKNOWN:     SENSOR_EXCEPTIONS.SENSOR_TIME_UNKNOWN_STR,
    SENSOR_EXCEPTIONS.SENSOR_LOST:             SENSOR_EXCEPTIONS.SENSOR_LOST_STR
}

is_connected = False
lastBolusTime = None
cycleTimer = None
cycleCount = 0

blynk = None
nightscout = None

CONFIG_FILE = "/etc/ddguard.conf"


def to_int(string):
   try:
      i = int(string)
   except:
      i = 0
   return i


#########################################################
#
# Function:    read_config()
# Description: Read parameters from config file
# 
#########################################################
def read_config(cfilename):
   
   # Parameters from global config file
   config = ConfigParser()
   config.read(cfilename)
   
   #TODO: check if file exists

   try:
      # Read Blynk parameters
      read_config.blynk_server    = config.get('blynk', 'server').split("#")[0].strip('"').strip("'").strip()
      read_config.blynk_token     = config.get('blynk', 'token').split("#")[0].strip('"').strip("'").strip()
      read_config.blynk_heartbeat = to_int(config.get('blynk', 'heartbeat').split("#")[0].strip('"').strip("'"))
   except ConfigParser.NoOptionError as NoSectionError:
      syslog.syslog(syslog.LOG_ERR, "ERROR - Needed blynk option not found in config file")
      return False

   try:
      # Read Nightscout parameters
      read_config.nightscout_server     = config.get('nightscout', 'server').split("#")[0].strip('"').strip("'").strip()
      read_config.nightscout_api_secret = config.get('nightscout', 'api_secret').split("#")[0].strip('"').strip("'").strip()
   except ConfigParser.NoOptionError as NoSectionError:
      syslog.syslog(syslog.LOG_ERR, "ERROR - Needed nightscout option not found in config file")
      return False

   try:
      # Read BGL alert parameters
      read_config.bgl_low_val      = to_int(config.get('bgl', 'bgl_low').split("#")[0].strip('"').strip("'"))
      read_config.bgl_pre_low_val  = to_int(config.get('bgl', 'bgl_pre_low').split("#")[0].strip('"').strip("'"))
      read_config.bgl_pre_high_val = to_int(config.get('bgl', 'bgl_pre_high').split("#")[0].strip('"').strip("'"))
      read_config.bgl_high_val     = to_int(config.get('bgl', 'bgl_high').split("#")[0].strip('"').strip("'"))
   except ConfigParser.NoOptionError as NoSectionError:
      syslog.syslog(syslog.LOG_ERR, "ERROR - Needed bgl option not found in config file")
      return False

   # Disable BGL parameters if not specified in config
   if read_config.bgl_pre_high_val == 0:
      read_config.bgl_pre_high_val = 1000
   if read_config.bgl_high_val == 0:
      read_config.bgl_high_val = 1000
      
   print ("Blynk server:    %s" % read_config.blynk_server)
   print ("Blynk token:     %s" % read_config.blynk_token)
   print ("Blynk heartbeat: %d\n" % read_config.blynk_heartbeat)
   print ("Nightscout server:     %s" % read_config.nightscout_server)
   print ("Nightscout api_secret: %s\n" % read_config.nightscout_api_secret)
   print ("BGL low:      %d" % read_config.bgl_low_val)
   print ("BGL pre low:  %d" % read_config.bgl_pre_low_val)
   print ("BGL pre high: %d" % read_config.bgl_pre_high_val)
   print ("BGL high:     %d\n" % read_config.bgl_high_val)
   return True

    
#########################################################
#
# Function:    on_sigterm()
# Description: signal handler for the TERM and INT signal
# 
#########################################################
def on_sigterm(signum, frame):
   try:
      if blynk != None:
         blynk.disconnect()
   except:
      pass
   try:
      if cycleTimer != None:
         cycleTimer.cancel()
   except:
      pass
   syslog.syslog(syslog.LOG_NOTICE, "Exiting DD-Guard daemon")
   sys.exit()


#########################################################
#
# Function:    blynk_upload()
# Description: Blynk uploader
# 
#########################################################
def blynk_upload(data):

   global lastBolusTime
   global cycleCount
   
   if data != None:
      print("Uploading data to Blynk")
       
      # Send sensor data
      if data["sensorBGL"] in sensor_exception_codes:
         # Sensor exception occured
         
         # BGL gauge
         blynk.virtual_write(VPIN_SENSOR, None)
         blynk.set_property(VPIN_SENSOR, "color", BLYNK_WHITE)
         
         # Trend and active insulin
         blynk.virtual_write(VPIN_ARROWS, "--"+" / "+str(data["activeInsulin"]))
         
         # Status line
         blynk.virtual_write(VPIN_STATUS, datetime.datetime.now().strftime("%H:%M")+" - "+sensor_exception_codes[data["sensorBGL"]])
         blynk.set_property(VPIN_STATUS, "color", BLYNK_RED)
      else:
         # Regular BGL data
         
         # BLG gauge
         blynk.virtual_write(VPIN_SENSOR, data["sensorBGL"])
         if data["pumpAlert"]["alertSuspend"] or data["pumpAlert"]["alertSuspendLow"]:
            blynk.set_property(VPIN_SENSOR, "color", BLYNK_BLUE)
         elif data["sensorBGL"] < read_config.bgl_low_val or data["sensorBGL"] > read_config.bgl_high_val or \
              data["pumpAlert"]["alertOnLow"] or data["pumpAlert"]["alertOnHigh"]:
            blynk.set_property(VPIN_SENSOR, "color", BLYNK_RED)
         elif data["sensorBGL"] < read_config.bgl_pre_low_val or data["sensorBGL"] > read_config.bgl_pre_high_val or \
              data["pumpAlert"]["alertBeforeLow"] or data["pumpAlert"]["alertBeforeHigh"]:
            blynk.set_property(VPIN_SENSOR, "color", BLYNK_YELLOW)
         else:
            blynk.set_property(VPIN_SENSOR, "color", BLYNK_GREEN)
         
         # Trend and active insulin
         blynk.virtual_write(VPIN_ARROWS, str(data["trendArrow"])+" / "+str(data["activeInsulin"]))
         
         # Status line
         calTime = "Cal at {0}".format((data["sensorBGLTimestamp"] + datetime.timedelta(minutes=data["sensorCalMinutesRemaining"])).strftime("%H:%M"))
         blynk.virtual_write(VPIN_STATUS, "Updated "+data["sensorBGLTimestamp"].strftime("%H:%M")+" - "+calTime)
         blynk.set_property(VPIN_STATUS, "color", BLYNK_GREEN)
       
      # Send pump data

      # Battery bar
      # Alternate pump and sensor battery
      if cycleCount%2 == 0:
         data_batt = data["batteryLevelPercentage"]
         label_batt = "PUMP BATTERY %"
      else:
         data_batt = data["sensorBatteryLevelPercentage"]
         label_batt = "SENSOR BATTERY %"
      blynk.set_property(VPIN_BATTERY, "label", label_batt)
      blynk.virtual_write(VPIN_BATTERY, data_batt)
      if data_batt <= 25:
         blynk.set_property(VPIN_BATTERY, "color", BLYNK_RED)
      elif data_batt <= 50:
         blynk.set_property(VPIN_BATTERY, "color", BLYNK_YELLOW)
      else:
         blynk.set_property(VPIN_BATTERY, "color", BLYNK_GREEN)
      
      # Reservoir bar
      blynk.virtual_write(VPIN_UNITS, int(round(data["insulinUnitsRemaining"])))
      if data["insulinUnitsRemaining"] <= 25:
         blynk.set_property(VPIN_UNITS, "color", BLYNK_RED)
      elif data["insulinUnitsRemaining"] <= 75:
         blynk.set_property(VPIN_UNITS, "color", BLYNK_YELLOW)
      else:
         blynk.set_property(VPIN_UNITS, "color", BLYNK_GREEN)
         
      # Active insulin / last bolus graph
      if int(data["lastBolusTime"].strftime("%s")) != lastBolusTime: 
         print("Bolus time changed")
         lastBolusTime = int(data["lastBolusTime"].strftime("%s"))
         # Check if last bolus time is recent
         if int(time.time()) - lastBolusTime < 2*UPDATE_INTERVAL:
            print("Bolus time is recent")
            blynk.virtual_write(VPIN_LASTBOLUS, data["lastBolusAmount"])
      else:
         blynk.virtual_write(VPIN_ACTINS, data["activeInsulin"])
      
   else:
      syslog.syslog(syslog.LOG_ERR, "Unable to get data from pump")
      blynk.set_property(VPIN_STATUS, "color", BLYNK_RED)


#########################################################
#
# Function:    upload_live_data()
# Description: Read live data from pump and upload it 
#              to the enabled cloud services
#              This runs once at startup and then as a 
#              periodic timer every 5min
# 
#########################################################
def upload_live_data():
   
   global cycleCount
   global cycleTimer
      
   # Guard against multiple threads
   if upload_live_data.active:
      return
    
   upload_live_data.active = True
   
   print("read live data from pump")
   hasFailed = True
   numRetries = MAX_RETRIES_AT_FAILURE
   while hasFailed and numRetries > 0:
      try:
         liveData = cnl24driverlib.readLiveData()
         hasFailed = False
      except:
         print("unexpected ERROR occured while reading live data")
         syslog.syslog(syslog.LOG_ERR, "Unexpected ERROR occured while reading live data")
         liveData = None
         numRetries -= 1
         if numRetries > 0:
            time.sleep(RETRY_DELAY)
            
   # Account for pump RTC drift
   if liveData != None:
      print("account for pump RTC drift:")
      print("   before: pumpTime {0},  sensorBGLTimestamp {1}".format(liveData["pumpTime"], liveData["sensorBGLTimestamp"]))
      liveData["pumpTime"] += liveData["pumpTimeDrift"]
      if liveData["sensorBGL"] != SENSOR_EXCEPTIONS.SENSOR_LOST:
         liveData["sensorBGLTimestamp"] += liveData["pumpTimeDrift"]
      print("   after : pumpTime {0},  sensorBGLTimestamp {1}".format(liveData["pumpTime"], liveData["sensorBGLTimestamp"]))
    
   # Upload data to Blynk server
   if blynk != None:
      try:
         blynk_upload(liveData)
      except:
         syslog.syslog(syslog.LOG_ERR, "Blynk upload ERROR")

   # Upload data to Nighscout server
   if nightscout != None:
      try:
         nightscout.upload(liveData)
      except:
         syslog.syslog(syslog.LOG_ERR, "Nightscout upload ERROR")
   
   # Calculate time until next reading
   if liveData != None:
      nextReading = liveData["sensorBGLTimestamp"] + datetime.timedelta(seconds=UPDATE_INTERVAL)
      tmoSeconds  = int((nextReading - datetime.datetime.now(liveData["pumpTime"].tzinfo)).total_seconds())
      print("Next reading at {0}, {1} seconds from now\n".format(nextReading,tmoSeconds))
      if tmoSeconds < 0:
         tmoSeconds = RETRY_INTERVAL
   else:
      tmoSeconds = RETRY_INTERVAL
      print("Retry reading {0} seconds from now\n".format(tmoSeconds))
      
   # Start timer for next cycle
   cycleTimer = threading.Timer(tmoSeconds+10, upload_live_data)
   cycleTimer.start()
   
   cycleCount += 1
   upload_live_data.active = False


##########################################################           
# Setup
##########################################################           

# read configuration parameters
if read_config(CONFIG_FILE) == False:
   sys.exit()

blynk_enabled = (read_config.blynk_token != "") and (read_config.blynk_server != "")
nightscout_enabled = (read_config.nightscout_server != "") and (read_config.nightscout_api_secret != "")

# Init Blynk instance
if blynk_enabled:
   print("Blynk upload is enabled")
   blynk = blynklib.Blynk(read_config.blynk_token,
                          server=read_config.blynk_server.strip(),
                          heartbeat=read_config.blynk_heartbeat)

   @blynk.handle_event("connect")
   def connect_handler():
      global is_connected
      if not is_connected:
         is_connected = True
         print("Connected to cloud server")
         syslog.syslog(syslog.LOG_NOTICE, "Connected to cloud server")

   @blynk.handle_event("disconnect")
   def disconnect_handler():
      global is_connected
      if is_connected:
         is_connected = False
         print("Disconnected from cloud server")
         syslog.syslog(syslog.LOG_NOTICE, "Disconnected from cloud server")

# Init Nighscout instance (if requested)
if nightscout_enabled:
   print("Nightscout upload is enabled")
   nightscout = nightscoutlib.nightscout_uploader(server = read_config.nightscout_server, 
                                                  secret = read_config.nightscout_api_secret)


##########################################################           
# Initialization
##########################################################           
syslog.syslog(syslog.LOG_NOTICE, "Starting DD-Guard daemon, version "+VERSION)

# Init signal handler
signal.signal(signal.SIGINT, on_sigterm)
signal.signal(signal.SIGTERM, on_sigterm)

upload_live_data.active = False

# Perform first upload immediately
# Subsequent uploads will be scheduled according to received data timestamp
t = threading.Thread(target=upload_live_data, args=())
t.start()

##########################################################           
# Main loop
##########################################################           
while True:
   if blynk_enabled:
      blynk.run()
   else:
      time.sleep(0.1)
